lib/i18n/tasks/deepl_translation.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+require 'deepl'
+require 'i18n/tasks/html_keys'
+
+module I18n::Tasks
+  module DeeplTranslation
+    # @param [I18n::Tasks::Tree::Siblings] forest to translate to the locales of its root nodes
+    # @param [String] from locale
+    # @return [I18n::Tasks::Tree::Siblings] translated forest
+    def deepl_translate_forest(forest, from)
+      forest.inject empty_forest do |result, root|
+        translated = translate_list(root.key_values(root: true), to: root.key, from: from)
+        result.merge! Data::Tree::Siblings.from_flat_pairs(translated)
+      end
+    end
+
+    # @param [Array<[String, Object]>] list of key-value pairs
+    # @return [Array<[String, Object]>] translated list
+    def translate_list(list, opts) # rubocop:disable Metrics/AbcSize
+      return [] if list.empty?
+      opts = opts.dup
+      opts[:key] ||= translation_config[:deepl_api_key]
+      validate_translate_api_key! opts[:key]
+      key_pos = list.each_with_index.inject({}) { |idx, ((k, _v), i)| idx.update(k => i) }
+      # copy reference keys as is, instead of translating
+      reference_key_vals = list.select { |_k, v| v.is_a? Symbol } || []
+      list -= reference_key_vals
+      result = list.group_by { |k_v| html_key? k_v[0], opts[:from] }.map do |is_html, list_slice|
+        fetch_translations list_slice, opts.merge(is_html ? { tag_handling: 'xml' } : { preserve_formatting: true })
+      end.reduce(:+) || []
+      result.concat(reference_key_vals)
+      result.sort! { |a, b| key_pos[a[0]] <=> key_pos[b[0]] }
+      result
+    end
+
+    # @param [Array<[String, Object]>] list of key-value pairs
+    # @return [Array<[String, Object]>] translated list
+    def fetch_translations(list, opts)
+      options = {
+        ignore_tags: %w[i18n]
```

Comment:
Does `ignore_tags` work even without `tag_handling: 'xml'`? 
If not, we may need to use the same trick as in the Google Translate implementation

Reply:
Yes, `ignore_tags` works even without other params. And should be fixed, because `%{user}` interpolation can be with `_html` / `.html` keys and on simple text tags.

📝 https://www.deepl.com/api.html